ODLA/platforms/dnnl/odla_dnnl.cc

```diff
@@ -507,28 +507,47 @@ odla_value odla_Cast(odla_value input, odla_element_type target_type,
   return CreateValue(ret_mem, input->shape, id);
 }
 
-static odla_value binary_eltwise_s32(dnnl::algorithm alg, odla_value lhs,
-                                     odla_value rhs, const odla_value_id id) {
+static odla_value binary_eltwise_s32(dnnl::algorithm alg, dnnl::memory lhs_mem,
+                                     dnnl::memory rhs_mem,
+                                     odla_value_shape shape,
+                                     const odla_value_id id) {
+  std::function<void()> elem_op;
+  int ln = GetTotalElements(shape);
+  elem_op = [lhs_mem, rhs_mem, ln, alg]() {
+    int32_t* rhs_ptr = nullptr;
+    rhs_ptr = (int32_t*)rhs_mem.get_data_handle();
+    dnnl_utils::binary_s32_func(alg, (int32_t*)lhs_mem.get_data_handle(),
+                                rhs_ptr, rhs_ptr, ln);
+  };
+  odla_value v = CreateValue(rhs_mem, shape, id);
+  add_op(elem_op);
+  return v;
+}
+static odla_value binary_eltwise(dnnl::algorithm algo, odla_value lhs,
+                                 odla_value rhs, const odla_value_id id) {
   const auto& dims_lhs = lhs->shape;
   const auto& dims_rhs = rhs->shape;
-  auto lhs_md =
-      dnnl::memory::desc(getDims(dims_lhs), lhs->mem.get_desc().data_type(),
-                         getFormatTag(dims_lhs));
+  auto lhs_md = dnnl::memory::desc(
+      getDims(dims_lhs), lhs->mem.get_desc().data_type(), getStrides(dims_lhs));
   auto ret_md = lhs_md;
+  auto rhs_md = lhs_md;
   auto ret_mem = dnnl::memory(ret_md, g_comp->eng);
-  auto rhs_md = rhs->mem.get_desc();
+  auto rhs_mem = dnnl::memory(rhs_md, g_comp->eng, rhs->mem.get_data_handle());
 
   auto ln = GetTotalElements(dims_lhs);
   auto rn = GetTotalElements(dims_rhs);
-  bool needs_broadcast = false;
+  bool broadcasted = false;
+  // do broadcast
   if (ln != rn) {
+    broadcasted = true;
     std::vector<int64_t> strides_v(dims_lhs.size, 0);
     assert(ln >= rn && ln % rn == 0);
     std::function<void()> op;
     if (rn == 1) {
-      op = [rhs, ret_mem, ln]() {
-        auto val = (int32_t*)rhs->mem.get_data_handle();
+      op = [rhs_mem, ret_mem, ln]() {
+        auto val = (int32_t*)rhs_mem.get_data_handle();
         memset(ret_mem.get_data_handle(), val[0], sizeof(int) * ln);
```

Comment:
Could you please double check with "memset"? It only uses one byte from src to populate the "dst" memory